packages/solid-query/src/createBaseQuery.ts

```diff
@@ -45,40 +49,100 @@ export function createBaseQuery<
+    })
+  }
+
+  const createClientSubscriber = (refetch: (info?: unknown) => void) => {
+    return observer.subscribe((result) => {
+      notifyManager.batchCalls(() => {
+        const unwrappedResult = { ...unwrap(result) }
+        setState(unwrappedResult)
+        refetch()
+      })()
+    })
+  }
+
+  /**
+   * Unsubscribe is set lazily, so that we can subscribe after hydration when needed.
+   */
+  let unsubscribe: (() => void) | null = null
+
+  const [queryResource, { refetch }] = createResource<
+    QueryObserverResult<TData, TError> | undefined
+  >(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          unsubscribe = createServerSubscriber(resolve)
+        } else {
+          if (!unsubscribe) {
+            unsubscribe = createClientSubscriber((info?: unknown) =>
+              refetch(info),
+            )
+          }
+
+          if (!state.isInitialLoading) {
+            resolve(state)
           }
-          resolve(unwrap(state.data))
         }
       })
     },
-  )
+    {
+      initialValue: state,
 
-  batch(() => {
-    mutate(() => unwrap(state.data))
-    refetch()
-  })
+      get deferStream() {
+        return options().deferStream
+      },
 
-  const unsubscribe = observer.subscribe((result) => {
-    notifyManager.batchCalls(() => {
-      const unwrappedResult = { ...unwrap(result) }
-      if (unwrappedResult.data === undefined) {
-        // This is a hack to prevent Solid
-        // from deleting the data property when it is `undefined`
-        // ref: https://www.solidjs.com/docs/latest/api#updating-stores
-        unwrappedResult.data = emptyData as any as undefined
-      }
-      setState(unwrap(unwrappedResult))
-      mutate(() => unwrap(result.data))
-      refetch()
-    })()
-  })
+      /**
+       * If this resource was populated on the server (either sync render, or streamed in over time), onHydrated
+       * will be called. This is the point at which we can hydrate the query cache state, and setup the query subscriber.
+       *
+       * Leveraging onHydrated allows us to plug into the async and streaming support that solidjs resources already support.
+       *
+       * Note that this is only invoked on the client, for queries that were originally run on the server.
+       */
+      onHydrated(_k, info) {
+        if (info.value) {
+          setState(info.value)
+          hydrate(queryClient(), {
+            queries: [
+              {
+                queryKey: defaultedOptions.queryKey,
+                queryHash: defaultedOptions.queryHash,
+                state: info.value,
+              },
+            ],
+          })
+        }
 
-  onCleanup(() => unsubscribe())
+        if (!unsubscribe) {
+          unsubscribe = createClientSubscriber((i?: unknown) => refetch(i))
```

Comment:
```suggestion
          unsubscribe = createClientSubscriber((i?: unknown) => refetch(i), true)
```
This would change to a true flag for hydrated queries only